### PR TITLE
LOOKUP JOIN using field-caps for field mapping

### PR DESCRIPTION
**Pull request by craigtaverner**
Removes the hard-coded hack for languages_lookup, and instead does a field-caps check for the real join index.

This is based on the initial work done in https://github.com/elastic/elasticsearch/pull/116515

For background look at the meta-task for `LOOKUP JOIN` at https://github.com/elastic/elasticsearch/issues/116208.

So basically think of it as a JOIN, and internally it is coded similarly to ENRICH. What we do for ENRICH, however, is get the field mappings by reading the enrich-policy. For the JOIN we have no such thing and need to get the field mappings from field_caps. When the query has a JOIN, there will be two `UnresolvedRelation` instances, one coming from the `FROM mainindex` clause and supporting a bunch of advanced stuff like wildcards and CCS, and the second coming from the `LOOKUP JOIN otherindex ON joinfield` call, and supporting much less (just single index name, no wildcards and no CCS). So we need to support three independent field mappings mechanisms:
* `ENRICH` (asynchronously read the enrich-policy)
* `LOOKUP JOIN` (asynchronously call field-caps on the `otherindex` with potentially needed field names, the join field name, as well as whatever the query is expecting to see later in the results)
* `FROM` (asynchronously call field-caps on the `mainindex`  with potentially needed field names, from the query plan. The index name could have wildcards and CCS support)

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java [hunk 1]
```diff
@@ -299,63 +304,81 @@ private <T> void preAnalyze(
...
-        LogicalPlan parsed,
+        List<TableInfo> indices,
         EsqlExecutionInfo executionInfo,
         Map<String, Exception> unavailableClusters,  // known to be unavailable from the enrich policy API call
-        ActionListener<IndexResolution> listener,
-        Set<String> enrichPolicyMatchFields
+        Set<String> fieldNames,
+        ActionListener<IndexResolution> listener
     ) {
-        PreAnalyzer.PreAnalysis preAnalysis = new PreAnalyzer().preAnalyze(parsed);
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:350]: :+1:

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java [hunk 2]
```diff
@@ -299,63 +304,81 @@ private <T> void preAnalyze(
...
+                        })
+                    )
+                )
+            );
         }));
     }
 
     private void preAnalyzeIndices(
-        LogicalPlan parsed,
+        List<TableInfo> indices,
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:369]: :+1:

**Review by nik9000 — APPROVED** [review]
LGTM. Wants a review from @astefan because he's refactoring this code and can probably answer questions like "TODO: why is this empty". If we can't get through all of the outstanding TODOs but @astefan is still happy let's get this merged because it's way better than my hack. We should just make sure to lodge the TODOs into the meta issue.

**Review by bpintea — COMMENTED** [review]
Just some fly-by questions.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java [hunk 3]
```diff
@@ -440,6 +483,11 @@ static Set<String> fieldNames(LogicalPlan parsed, Set<String> enrichPolicyMatchF
                 // The exact name of the field will be added later as part of enrichPolicyMatchFields Set
                 enrichRefs.removeIf(attr -> attr instanceof EmptyAttribute);
                 references.addAll(enrichRefs);
+            } else if (p instanceof LookupJoin join) {
+                keepJoinReferences.addAll(join.config().matchFields());  // TODO: why is this empty
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:487]: Had to have a look: the LookupJoin is [created with empty join fields](https://github.com/elastic/elasticsearch/blob/041022944e7df963d4a3269dfa6ec7791247251e/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/LookupJoin.java#L32), and the join fields are extracted from the `UsingJoinType` only later in the optimiser (when `surrogate()` is invoked) -- I suppose using the empty fields needs fixing (at some point before or after this PR?).

**Comment by bpintea** [comment]
Some tests might help -- using an inexistent join field isn't caught by the verifier (but still caught, though too late, in `LogicalVerifier`).

**Comment by nik9000** [comment]
Indeed, more tests is probably good. I'd personally use the yaml stuff to build different lookup indices scenarios and try them.

**Comment by nik9000** [comment]
I guess you'd be the first to make YAML tests for this.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/AnalyzerContext.java [hunk 4]
```diff
@@ -15,5 +15,15 @@ public record AnalyzerContext(
     Configuration configuration,
     EsqlFunctionRegistry functionRegistry,
     IndexResolution indexResolution,
+    IndexResolution lookupResolution,
     EnrichResolution enrichResolution
-) {}
+) {
+    public AnalyzerContext(
```
[the remark this answers is not in the document]
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/AnalyzerContext.java:21]: It was for tests, but I did not want to follow the test approach to enrichResolution, which has three different ways of passing in an empty resolution, cluttering the tests. So this is an intermediate solution, which minimizes code changes (does not change test code), while getting the goal of the PR achieved. Then I would like a second PR that simplifies both this and the enrich resolution for tests (and cleans up test code all over the place).

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java [hunk 5]
```diff
@@ -299,63 +304,81 @@ private <T> void preAnalyze(
...
+                                    enrichPolicyResolver.resolvePolicies(
+                                        newClusters,
+                                        unresolvedPolicies,
+                                        ll.map(
+                                            newEnrichResolution -> action.apply(indexResolution, lookupIndexResolution, newEnrichResolution)
+                                        )
+                                    );
+                                    return;
+                                }
+                            }
```
[the remark this answers is not in the document]
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:359]: Agreed. But I think that should be done separately. I've tried to keep the CCS code as similar as possible to what was there before to reduce conflicts with the people working on CCS.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java [hunk 6]
```diff
@@ -401,6 +424,25 @@ private void preAnalyzeIndices(
         }
     }
 
+    private void preAnalyzeLookupIndices(List<TableInfo> indices, Set<String> fieldNames, ActionListener<IndexResolution> listener) {
+        if (indices.size() > 1) {
```
[the remark this answers is not in the document]
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:428]: I believe the overall project plan has steps for improving this, and it requires changes in many more places. This PR focuses on keeping the same minimal test set working, while simply replacing the hard-coded index mappings with actual field-caps calls. I think a followup PR should deal with multiple index mappings (and multiple field-caps calls).

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java [hunk 7]
```diff
@@ -200,56 +200,44 @@ private static class ResolveTable extends ParameterizedAnalyzerRule<UnresolvedRe
         @Override
         protected LogicalPlan rule(UnresolvedRelation plan, AnalyzerContext context) {
             if (plan.indexMode().equals(IndexMode.LOOKUP)) {
-                return hackLookupMapping(plan);
+                return resolveIndex(plan, context.lookupResolution());
             }
-            if (context.indexResolution().isValid() == false) {
-                return plan.unresolvedMessage().equals(context.indexResolution().toString())
+            return resolveIndex(plan, context.indexResolution());
```
[the remark this answers is not in the document]
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:205]: OK. I've made that simplification. My reason for the original version was I thought it looked more like the old code, so might be easier to review.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java (hunk 3 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:487]: After this PR, hence the TODO to remind us to fix this. I view this work as a first minimal fix to the hard-coded field mappings, and want it to be clean, simple and easy to review.

**Comment by craigtaverner** [comment]
> Indeed, more tests is probably good. I'd personally use the yaml stuff to build different lookup indices scenarios and try them.

I had hoped that this PR would simply fix the hard-coded mapping, making sure the existing csv-spec tests still worked. Increasing the tests will no-doubt expose quite a few serious limitations in the current stack for LOOKUP JOIN, which I assumed would be for a followup PR to tackle. If we do that in this PR, then we should do it because we do want this PR to be a much broader PR. My vote is lets make 2 PRs, this one does only what it claims in the description, and a followup that broadens the scope of the LOOKUP JOIN to handle multiple joins, and a broader suite of test cases. I can start on that directly, but would like this merged first.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:359]: Indeed, I hope I could clean up the EsqlSession code a bit with https://github.com/elastic/elasticsearch/pull/116755.

**Comment by nik9000** [comment]
> Increasing the tests will no-doubt expose quite a few serious limitations in the current stack for LOOKUP JOIN

That's fine with me.

**Review by astefan — APPROVED** [review]
LGTM 👍

**Review by costin — APPROVED** [review]
👍

**Comment by alex-spies** [comment]
This is missing a backport to 8.x.

**Comment by astefan** [comment]
The missing backport is blocking https://github.com/elastic/elasticsearch/pull/116755 backport.

**Comment by nik9000** [comment]
Looks like backport is coming in https://github.com/elastic/elasticsearch/pull/117967

**Pull request merged**